pavelo8501/Smart-Downloader
# Download service

Summary
This pull request introduces a robust FileDownloadService class designed for modular and scalable file download operations. The new implementation incorporates the following enhancements:

Modular Design:

Separation of concerns via a DownloadConnectorInterface, making the file download process flexible and extendable.
Support for both single-part and multipart downloads.

Callback Integration:
reportStatus callback for real-time status updates (e.g., start, complete).
handleProgress callback for granular tracking of download progress.

Resume Capability (Partial Stub):
The groundwork for download resumption is laid, with future implementation pending database integration.

Error Handling:
Graceful management of download interruptions, exceptions, and connection issues.

Key Changes
FileDownloadService Class:

Added methods:
start(string $url, int $chunk_size, TransactionDataClass $transaction): Initiates a file download.
stop(string $message = ""): Stops the download process with an optional message.
resume(string $url, int $chunkSize, int $byteOffset): Placeholder method for download resumption.
Integrated DownloadConnectorInterface for improved modularity.

CurlServiceConnector Class:
Implements the DownloadConnectorInterface using cURL.
Handles both single-part and multipart downloads.
Introduces a headerLookup method to inspect the server's file details (e.g., size, ranges).

Interface Definition:
DownloadConnectorInterface defines the contract for any connector plugin:
downloadFile
headerLookup
stopDownload

DownloadDataClass Enhancements:
Extended to support progress tracking and byte offsets for multipart downloads.
Testing
Unit Tests:

Added PHPUnit tests for FileDownloadService to validate:
Proper initialization and triggering of downloads (testStart_ShouldTriggerDownload).
Callback invocation for progress and status reporting (testStart_ShouldReportStatusAndHandleProgress).
Stopping the download process (testStop_ShouldCallConnectorStopDownload).
Resume functionality placeholder test (testResume_NotImplementedYet).

Manual Tests:
Verified single-part and multipart downloads using CurlServiceConnector.
Pending Tasks

Framework integration (Laravel) to check design consistency:
Database Integration:
Logging Functionality:

Implement the resume method to leverage stored download state for resuming interrupted downloads.
Advanced Error Handling:

Improve user feedback for edge cases (e.g., invalid URLs, interrupted connections).
